src/interp/istream.h

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef WABT_INTERP_ISTREAM_H_
+#define WABT_INTERP_ISTREAM_H_
+
+#include <cstdint>
+#include <vector>
+#include <string>
+
+#include "src/common.h"
+#include "src/opcode.h"
+#include "src/stream.h"
+
+namespace wabt {
+namespace interp {
+
+using u8 = uint8_t;
+using u16 = uint16_t;
+using u32 = uint32_t;
+using u64 = uint64_t;
+using f32 = float;
+using f64 = double;
+
+using Buffer = std::vector<u8>;
+
+using ValueType = wabt::Type;
+
+enum class InstrKind {
+  Imm_0_Op_0,             // Nop
+  Imm_0_Op_1,             // i32.eqz
+  Imm_0_Op_2,             // i32.add
+  Imm_0_Op_3,             // select
+  Imm_Jump_Op_0,          // br
+  Imm_Jump_Op_1,          // br_if
+  Imm_Index_Op_0,         // global.get
+  Imm_Index_Op_1,         // global.set
+  Imm_Index_Op_2,         // table.set
+  Imm_Index_Op_3,         // memory.fill
+  Imm_Index_Op_N,         // call
+  Imm_Index_Index_Op_3,   // memory.init
+  Imm_Index_Index_Op_N,   // call_indirect
+  Imm_Index_Offset_Op_1,  // i32.load
+  Imm_Index_Offset_Op_2,  // i32.store
+  Imm_Index_Offset_Op_3,  // i32.atomic.rmw.cmpxchg
+  Imm_I32_Op_0,           // i32.const
+  Imm_I64_Op_0,           // i64.const
+  Imm_F32_Op_0,           // f32.const
+  Imm_F64_Op_0,           // f64.const
+  Imm_I32_I32_Op_0,       // drop_keep
+  Imm_I8_Op_1,            // i32x4.extract_lane
+  Imm_I8_Op_2,            // i32x4.replace_lane
+  Imm_V128_Op_0,          // v128.const
+  Imm_V128_Op_2,          // v8x16.shuffle
+};
+
+struct Instr {
+  Opcode op;
+  InstrKind kind;
+  union {
+    u8 imm_u8;
+    u32 imm_u32;
+    f32 imm_f32;
+    u64 imm_u64;
+    f64 imm_f64;
+    v128 imm_v128;
+    struct { u32 fst, snd; } imm_u32x2;
+  };
+};
+
+struct Istream {
```

Comment:
I would have though something like this with non-trivial methods and private fields would be a class?  Do you have set of rules for when you choose class vs struct?

Reply:
No, I should probably be more consistent about this. You're right that this should be a class.